### PR TITLE
Enhance/error handling

### PR DESCRIPTION
**Pull request by mhmd-android**
**Error Handling Improvement**

>Hi, I'm back with a new PR :). This PR primarily aims to enhance error handling across the project, but it also introduces significant improvements and refactoring to the overall code structure.

One of the most significant improvements in this PR is the introduction of a robust and flexible API response handling system using Ktor. This new system provides a more structured and type-safe way to manage API responses and errors.

**Key Components:**

**1. ApiResponse Sealed Interface:**
- This is the core of our new response handling system.
- It defines three main types of responses: a. `Success<T>`: Represents successful API calls with data. b. `Failure.Error`: Represents API errors (e.g., server errors). c. `Failure.Exception`: Represents client-side exceptions (e.g., network issues).

**2. apiResponseOf Function:**
- A suspend function that wraps API calls and returns an `ApiResponse`.
- Automatically categorizes responses based on status codes.
- Catches exceptions and wraps them in `ApiResponse.Failure.Exception`

**3. Extension Functions for HttpResponse:**
- `getStatusCode()`: Maps HTTP status codes to our custom StatusCode enum.
- Properties like `payloadResponse`and `statusCode `for easier access to error details.

**4. HttpClient Extensions:**
-  `getApiResponse`: Simplifies making GET requests and handling responses.

**5. ApiResponse Extension Functions:**
- A set of functions like `onSuccess`, `onError`, `onException`, etc.
- Allows for more functional and readable API response handling.
- Includes both synchronous and suspend versions for flexibility.

**6. Error Handling:**

- Created a new Errors sealed class to categorize different types of errors.
- Updated `Resource` class to use the new `Errors` type
- Implemented error mapping in data layer to convert exceptions to appropriate `Errors` types.

**Benefits of This Approach:**

1. **Type Safety:** The sealed interface ensures that all possible response types are handled.
2. **Separation of Concerns:** Clearly separates successful responses, API errors, and exceptions.
3. **Flexibility:** The extension functions allow for easy chaining of response handling logic.
4. **Coroutine Support:** Designed to work seamlessly with Kotlin coroutines.
5. **Improved Error Handling:** Provides detailed information about different types of failures.

**One of Usage Example::**
```
client.getApiResponse<UserData> {
    url("https://api.example.com/user")
}.onSuccess { 
    // Handle successful response
}.onError { 
    // Handle API errors
}.onException {
    // Handle exceptions
}
```

> This new system replaces the previous approach of handling Ktor responses with individual functions, providing a more cohesive and maintainable solution. It significantly improves error handling capabilities and makes API interactions more predictable and easier to manage throughout the application.

**Important Note:**
- Despite substantial improvements, more work is needed on the UI side for proper error display.

Thank you for your review and feedback.
#205

### CODE REVIEW

File: core/ktor/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Kaaveh) [inline core/ktor/src/main/AndroidManifest.xml]: I think we can delete this file

File: core/ktor/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+plugins {
+    id("composenews.android.library")
+    libs.plugins.apply {
+        alias(kotlinx.serialization)
+    }
+}
+
+android {
+    namespace = "ir.composenews.ktor"
```
Comment (Kaaveh) [inline core/ktor/build.gradle.kts]: This is a module about networking in general. I think `ktor` is very specific for this purpose. For example, this library may change with another networking library. So, I propose renaming `core/ktor` to a more proper name.

File: core/ktor/src/main/java/ir/composenews/ktor/ResponseTransformer.kt [hunk 3]
```diff
@@ -0,0 +1,96 @@
+
+package ir.composenews.ktor
+
+import kotlin.contracts.ExperimentalContracts
```
Comment (Kaaveh) [inline core/ktor/src/main/java/ir/composenews/ktor/ResponseTransformer.kt]: Nice! Do you get inspiration from somewhere else or any article?

**Review by Kaaveh — CHANGES_REQUESTED** [review]

File: core/ktor/src/main/AndroidManifest.xml (hunk 1 above)
Comment (mhmd-android) [reply core/ktor/src/main/AndroidManifest.xml]: Certainly, I'll remove the file

File: core/ktor/build.gradle.kts (hunk 2 above)
Comment (mhmd-android) [reply core/ktor/build.gradle.kts]: I appreciate your insight. You're absolutely right, and I should have considered a more generic name from the start. I wasn't entirely sure about the naming when I began, but your point about potential future changes in networking libraries is spot-on. A more general name would indeed be more appropriate.
What do you think about renaming it to something like 'core-networking'? This would keep it generic while still indicating its purpose. Do you have any other suggestions for the module name that you think might fit better?

File: core/base/build.gradle.kts [hunk 4]
```diff
@@ -10,6 +10,7 @@ android {
 dependencies {
     projects.library.apply {
         api(projects.core.test)
+        api(projects.core.ktor)
```
Comment (VahidGarousi) [inline core/base/build.gradle.kts]: In my opinion we can use something like this:
```
projects.core.network.ktor
```

File: core/ktor/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+plugins {
+    id("composenews.android.library")
```
Comment (VahidGarousi) [inline core/ktor/build.gradle.kts]: Move this ```composenews.android.library``` to ```libs.versions.tom``` and plugin section like this:

```
composenew-android-library= { id = "composenews.android.library", version = "unspecified" }
```
and use it like this:
```
alias(libs.plugins.composenews.android.library)
```

File: core/ktor/src/main/java/ir/composenews/ktor/ApiResponse.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package ir.composenews.ktor
+
+sealed interface ApiResponse<out T> {
```
Comment (VahidGarousi) [inline core/ktor/src/main/java/ir/composenews/ktor/ApiResponse.kt]: Can you provide a proposal for this sealed interface?

What is your intention/motivation for adding this?

File: core/ktor/src/main/java/ir/composenews/ktor/ResponseTransformer.kt (hunk 3 above)
Comment (mhmd-android) [reply core/ktor/src/main/java/ir/composenews/ktor/ResponseTransformer.kt]: When I was researching API management in Ktor, I came across this library: [https://github.com/skydoves/sandwich](https://github.com/skydoves/sandwich)
Reading through its code was incredibly educational, and I felt we could draw inspiration from parts of this repo and customize it to fit our specific needs. While we could have added the entire library, my goal was to tailor it to our particular project and learn from it in the process. The Sandwich library is indeed very powerful, and customizing it for our use case not only served our project's needs but also provided a great learning opportunity. It allowed us to understand the intricacies of API response handling in Ktor while adapting the concepts to our specific requirements

File: core/ktor/src/main/java/ir/composenews/ktor/HttpClientExtension.kt [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package ir.composenews.ktor
+
+import io.ktor.client.HttpClient
+import io.ktor.client.request.HttpRequestBuilder
+import io.ktor.client.request.request
+import io.ktor.http.HttpMethod
+
+suspend inline fun <reified T> HttpClient.getApiResponse(
```
Comment (VahidGarousi) [inline core/ktor/src/main/java/ir/composenews/ktor/HttpClientExtension.kt]: If this method does ```GET``` request so we can rename it to ```get```

and we don't need ```ApiResponse``` postfix. because our context here is HttpClient and we are explicitly talking about GET request

File: core/ktor/src/main/java/ir/composenews/ktor/ResponseTransformer.kt (hunk 3 above)
Comment (VahidGarousi) [reply core/ktor/src/main/java/ir/composenews/ktor/ResponseTransformer.kt]: Thank you for your innovation 🙏 
Can you describe the benefit of this methods?

File: core/base/build.gradle.kts (hunk 4 above)
Comment (mhmd-android) [reply core/base/build.gradle.kts]: Thank you for sharing your suggestion. That's an good idea, and I agree with you. Using `projects.core.network.ktor` provides a clear and structured naming convention that reflects both the general purpose (network) and the specific implementation (ktor). This approach maintains clarity while allowing for potential future additions or changes. I appreciate your input on this matter.

File: core/ktor/src/main/java/ir/composenews/ktor/ApiResponse.kt (hunk 6 above)
Comment (mhmd-android) [reply core/ktor/src/main/java/ir/composenews/ktor/ApiResponse.kt]: Thank you for your question about the ApiResponse sealed interface. My intention and motivation for adding this interface are as follows:

Type-safe response handling: The sealed interface allows us to handle different types of API responses (success, error, exception) in a type-safe manner, reducing the chances of runtime errors.

Separation of concerns: By categorizing responses into distinct types, we can more easily separate the logic for handling successful responses from error handling.

Improved error handling: The Failure types (Error and Exception) allow us to distinguish between server-side errors and client-side exceptions, enabling more precise error handling and reporting

Consistency across the application: This structure provides a standardized way to handle API responses throughout the entire application, making the codebase more maintainable and easier to understand.

Easier testing: With clearly defined response types, it becomes simpler to write unit tests for different scenarios, including edge cases and error conditions.

Flexibility for future expansion: If we need to add more specific response types in the future, the sealed interface structure allows us to do so without breaking existing code.


The ApiResponse interface is inspired by best practices in modern Android development and libraries like Sandwich, but customized to fit our specific project needs. It's designed to enhance our ability to manage API responses effectively and robustly.

Does this explanation help clarify my intentions? I'm open to any suggestions or improvements you might have for this approach.

File: core/ktor/src/main/java/ir/composenews/ktor/HttpClientExtension.kt (hunk 7 above)
Comment (mhmd-android) [reply core/ktor/src/main/java/ir/composenews/ktor/HttpClientExtension.kt]: You're absolutely right. The suggestion is logical. Given that we're working within the context of HttpClient in Ktor, using the `ApiResponse` suffix seems redundant.

Also, this class can be easily expanded in the future to include more HTTP methods such as POST, PUT, and others, based on our project requirements

File: core/ktor/src/main/java/ir/composenews/ktor/ResponseTransformer.kt (hunk 3 above)
Comment (mhmd-android) [reply core/ktor/src/main/java/ir/composenews/ktor/ResponseTransformer.kt]: @VahidGarousi 

Certainly
These extension functions provide a powerful and flexible way to handle different states of the `ApiResponse`

Chaining capability:
Most of these functions return ApiResponse<T>, allowing you to chain multiple operations. This leads to more readable and concise code.

Suspend function support:
The 'suspend' versions of these functions allow for easy integration with coroutines

Mapping functionality:
The 'suspendMap' functions allow you to transform the content of Success, Error, or Exception responses, providing flexibility in how you process the data.

Let us have the advantages of these helper methods with an example:
Using `suspendOnSuccess`: Let's look at this
```
override suspend fun syncMarketList() {
    api.getMarkets(
        "usd",
        "market_cap_desc",
        20,
        1,
        false,
    ).suspendOnSuccess {
        val toRemoteMarketDto = data.map {
            it.toRemoteMarketDto()
        }
        dao.upsertMarket(toRemoteMarketDto)
    }.onError {
        Log.d("debug", message)
    }.onException {
        Log.d("debug", message.toString())
    }
}
```
Here, `suspendOnSuccess` is used because inside the success block, you're calling another suspend function: `dao.upsertMarket(toRemoteMarketDto)`. This is exactly the "Suspend function support" you mentioned.

> This approach allows you to perform other asynchronous operations if the API call is successful, without worrying about managing threads or callbacks.


Using `suspendMap`: Now, let's look at your second example:
```
override fun fetchChart(id: String): Flow<Resource<Chart, Errors>> = flow {
    val chart = api.getMarketChart(id, "usd", 1)
    chart.suspendOnSuccess {
        suspendMap {
            emit(Resource.Success(it.data.toChart()))
        }
    }.suspendOnError {
        suspendMap {
            emit(
                Resource.Error(
                    error = Errors.ApiError(
                        it.statusCode.mapMessageStatusCode(),
                        it.statusCode.code,
                    ),
                ),
            )
        }
    }.suspendOnException {
        suspendMap {
            emit(Resource.Error(Errors.ExceptionError(it.message, throwable)))
        }
    }
}
```
Here, `suspendMap` acts as a mapper, This method allows you to transform the API result as desired and then `emit` it in the `Flow`.

> The use of `suspendMap` in this context is particularly useful when you need to perform some asynchronous operations or transformations on your data before `emitting it to the Flow`. It seamlessly integrates with the coroutine-based approach, maintaining the asynchronous nature of your operations while providing a clean and expressive way to handle different response states.


Indeed, that's exactly why we have both `suspendOnSuccess` and `onSuccess` versions, as well as corresponding versions for error and exception handling. This flexibility allows us to handle different scenarios efficiently. Let's break it down:

1. Regular versions (`onSuccess`, `onError`, `onException`):
These are used when you don't need to perform any suspend operations in the callback. They're suitable for simple transformations or actions that don't require coroutine context.

2. Suspend versions (`suspendOnSuccess`, `suspendOnError`, `suspendOnException`):
These are used when you need to perform suspend operations within the callback. They're perfect for scenarios where you need to call other suspend functions or perform long-running operations without blocking the main thread.

3. `suspendMap` for each of these:
The suspendMap function acts as a transformer for each of these callbacks. It allows you to perform asynchronous transformations on your data before finalizing the result.

**Here's a conceptual overview of how these functions relate:**
```
// Regular versions
onSuccess { /* non-suspend operations */ }
onError { /* non-suspend error handling */ }
onException { /* non-suspend exception handling */ }

// Suspend versions
suspendOnSuccess { /* can include suspend operations */ }
suspendOnError { /* can include suspend operations for error handling */ }
suspendOnException { /* can include suspend operations for exception handling */ }

// suspendMap versions for transformation
suspendOnSuccess {
    suspendMap { /* asynchronous transformation */ }
}
suspendOnError {
    suspendMap { /* asynchronous error transformation */ }
}
suspendOnException {
    suspendMap { /* asynchronous exception transformation */ }
}
```

**In conclusion,
The Sandwich library is truly excellent and provides a comprehensive explanation of its functionality. Its design and implementation are really impressive. My goal in not directly adding the library, but rather referencing it and implementing our own version, was a smart strategy for several reasons:**

**By implementing it ourselves, we're forced to dive deep into the code and understand the underlying mechanisms. This approach ensures that we're not just using a tool, but truly understanding how it works. If we need to add or modify functionality in the future, we already have a solid understanding of the codebase. This makes it easier to extend or customize the functionality to suit our specific needs.**

**This approach encourages continuous learning and improvement. By regularly revisiting and potentially expanding on this implementation, we're keeping our skills sharp and up-to-date. By implementing it ourselves, we have full control and understanding of every part of our codebase, which can be crucial for debugging and optimization.**

**While it would have been easier to simply add the library directly, this approach of "forced engagement" with the code is beneficial. It reflects our commitment to learning and mastery of our craft.**

Thank you for taking the time to read through this detailed explanation🙏

File: core/ktor/src/main/java/ir/composenews/ktor/ApiResponse.kt (hunk 6 above)
Comment (VahidGarousi) [reply core/ktor/src/main/java/ir/composenews/ktor/ApiResponse.kt]: Thank you for the detailed and clear explanation of the ApiResponse. I appreciate the approach you've proposed, and your intention behind adding this interface is well understood.

I have a suggestion that might enhance the flexibility and error management further. I propose introducing an Error interface to decouple errors from the result structure, allowing us to manage errors more independently and categorize them more effectively.

Additionally, by introducing a new sealed interface called Result with Success and Error types, we can achieve more type-safety and use any combination of data and error types. This approach could simplify testing and make the codebase more maintainable.

Finally, adding utility functions like map and asEmptyDataResult can make working with results more fluent and improve the readability of the code.

Here's an example of what I’m suggesting:

```kotlin
interface Error

sealed interface DataError: Error {
    enum class Network: DataError {
        REQUEST_TIMEOUT,
        UNAUTHORIZED,
        CONFLICT,
        TOO_MANY_REQUESTS,
        NO_INTERNET,
        PAYLOAD_TOO_LARGE,
        SERVER_ERROR,
        SERIALIZATION,
        UNKNOWN
    }

    enum class Local: DataError {
        DISK_FULL
    }
}

sealed interface Result<out D, out E> {
    data class Success<out D>(val data: D) : Result<D, Nothing>
    data class Error<out E : composenews.core.domain.util.Error>(val error: E) : Result<Nothing, E>
}

inline fun <T, E : Error, R> Result<T, E>.map(map: (T) -> R): Result<R, E> {
    return when (this) {
        is Result.Error -> Result.Error(error)
        is Result.Success -> Result.Success(map(data))
    }
}

fun <T, E : Error> Result<T, E>.asEmptyDataResult(): EmptyResult<E> {
    return map { }
}

typealias EmptyResult<E> = Result<Unit, E>

```


I believe these suggestions could help strengthen the current structure. I’d love to hear your thoughts on this.

**Review by Kaaveh — APPROVED** [review]

**Comment by Kaaveh** [comment]
@VahidGarousi are we ready to merge the PR?

**Review by VahidGarousi — APPROVED** [review]

File: core/ktor/src/main/java/ir/composenews/ktor/ApiResponse.kt (hunk 6 above)
Comment (mhmd-android) [reply core/ktor/src/main/java/ir/composenews/ktor/ApiResponse.kt]: Thank you for your suggestion 🙏
and I'd like to ensure I fully grasp its implementation. I'm particularly interested in understanding if this new structure should be applied consistently across all layers of the application, or if there's a more nuanced implementation you had in mind.
If possible, a small code snippet demonstrating how this would look in practice would be immensely helpful. This would allow us to visualize how the Error interface and Result structure interact in real scenarios.
I greatly appreciate your suggestion my buddy, and I'm looking forward to learning more about how we can optimize error handling and data flow based on your recommendation.
